### PR TITLE
Improved version reporting in NCCL_DEBUG=VERSION

### PR DESCRIPTION
**Pull request by rahulvaidya20**
## Details
___Do not mention proprietary info or link to internal work items in this PR.___

**Work item:** Internal

**What were the changes?**  
Improves HIP and ROCm versioning info (includes build number, patch and githash). Additionally displays hostname and path to librccl.so.

**Why were the changes made?**  
Displaying complete version info helps in understanding the rocm/hip build type (mainline, staging etc). Enabling support for displaying hostname can retrieve platform info (dell, banff, etc if specified in hostname). Path to librccl.so is important to understand user's librccl path loaded by the dynamic linker at runtime.

**How was the outcome achieved?**  
**Changes to init.cc**: Added a new macro, VERSION_STRING_EXTENDED to display hip and rocm version. Added support to display hostname using **gethostname** function and librccl path using **dladdr** function.
**New file**: 
1. **hip_rocm_version_info.h**: Checks rocm version and sets the macros HIP_BUILD_INFO, ROCM_BUILD_INFO accordingly.

**Additional Documentation:**  

Sample output:

image

## Approval Checklist
___Do not approve until these items are satisfied.___
- [ ] Verify the CHANGELOG has been updated, if
  - there are any NCCL API version changes,
  - any changes impact library users, and/or
  - any changes impact any other ROCm library.

### CODE REVIEW

**Review by wenkaidu — APPROVED** [review]

File: src/init.cc [hunk 1]
```diff
@@ -683,17 +689,40 @@ static ncclResult_t devCommSetup(ncclComm_t comm) {
...
 #else
-#define VERSION_STRING "NCCL version " STR(NCCL_MAJOR) "." STR(NCCL_MINOR) "." STR(NCCL_PATCH) NCCL_SUFFIX "+cuda" STR(CUDA_MAJOR) "." STR(CUDA_MINOR)
+#define VERSION_STRING "NCCL version " STR(NCCL_MAJOR) "." STR(NCCL_MINOR) "." STR(NCCL_PATCH) NCCL_SUFFIX
+#define VERSION_STRING_EXTENDED "CUDA version " STR(CUDA_MAJOR) "." STR(CUDA_MINOR)
 #endif
 static void showVersion() {
   static int shown = 0;
   if (shown == 0 && ncclDebugLevel >= NCCL_LOG_VERSION) {
-    printf("%s %s\n", VERSION_STRING, rcclGitHash);
+    char hostInfo[512], libPathInfo[2048];
```
Comment (corey-derochie-amd) [inline src/init.cc]: I think you can use `HOST_NAME_MAX` for `hostInfo` size, I just need to see if that's portable to wherever we compile. See: https://www.man7.org/linux/man-pages/man2/gethostname.2.html

File: src/init.cc [hunk 2]
```diff
@@ -683,17 +689,40 @@ static ncclResult_t devCommSetup(ncclComm_t comm) {
...
+    
+    // Retireve Hostname info
+    if (gethostname(hostInfo, 512) != 0) {
+      // Returns Unknown in hostInfo if function call unsuccessful
+      strncpy(hostInfo, "Unknown", 512);
+    }
```
Comment (corey-derochie-amd) [inline src/init.cc]: I'd recommend `sizeof(hostInfo)` for a generic way.

Also, some systems return success when truncating, but it is unspecified whether the returned buffer includes a null terminator. So you may want to do -1 and add a terminator, or have a +1 sized array. See https://www.man7.org/linux/man-pages/man2/gethostname.2.html

File: src/init.cc [hunk 3]
```diff
@@ -683,17 +689,40 @@ static ncclResult_t devCommSetup(ncclComm_t comm) {
...
+    
+    // Retrieve librccl path
+    Dl_info pathInfo;
+    if (dladdr((void*)ncclCommInitRank, &pathInfo)) {
+      strncpy(libPathInfo, pathInfo.dli_fname, 2047);
+      libPathInfo[2047] = '\0'; //To prevent overflow
+    } else {
+      // Sets libPath to Unknown if the above function call is not successful
+      strncpy(libPathInfo, "Unknown", 2048);
```
Comment (corey-derochie-amd) [inline src/init.cc]: I'd recommend using `sizeof(libPathInfo)`.

**Review by corey-derochie-amd — APPROVED** [review]
LGTM, with some suggestions.

**Review by corey-derochie-amd — APPROVED** [review]
LGTM

**Review by nileshnegi — APPROVED** [review]

File: src/init.cc [hunk 4]
```diff
@@ -683,17 +689,44 @@ static ncclResult_t devCommSetup(ncclComm_t comm) {
...
 #endif
 static void showVersion() {
   static int shown = 0;
   if (shown == 0 && ncclDebugLevel >= NCCL_LOG_VERSION) {
-    printf("%s %s\n", VERSION_STRING, rcclGitHash);
+    char hostInfo[HOST_NAME_MAX], libPathInfo[2048];
+    size_t hostInfoSize = sizeof(hostInfo);
+    size_t libPathInfoSize = sizeof(libPathInfo);
+    
+    // Retireve Hostname info
```
Comment (gilbertlee-amd) [inline src/init.cc]: spelling error

File: src/init.cc [hunk 5]
```diff
@@ -683,17 +689,44 @@ static ncclResult_t devCommSetup(ncclComm_t comm) {
...
+    if (gethostname(hostInfo, hostInfoSize-1) == 0) {
+      hostInfo[hostInfoSize-1] = '\0'; //To prevent overflow
+    } else {
+      // Returns Unknown in hostInfo if function call unsuccessful
+      strncpy(hostInfo, "Unknown", hostInfoSize);
+    }
```
Comment (gilbertlee-amd) [inline src/init.cc:711]: Is there any particular reason to not just initialize hostInfo to {}, then pass in sizeof(hostInfo) to gethostname?

File: src/init.cc [hunk 6]
```diff
@@ -683,17 +689,44 @@ static ncclResult_t devCommSetup(ncclComm_t comm) {
...
   static int shown = 0;
   if (shown == 0 && ncclDebugLevel >= NCCL_LOG_VERSION) {
-    printf("%s %s\n", VERSION_STRING, rcclGitHash);
+    char hostInfo[HOST_NAME_MAX], libPathInfo[2048];
+    size_t hostInfoSize = sizeof(hostInfo);
+    size_t libPathInfoSize = sizeof(libPathInfo);
+    
+    // Retireve Hostname info
+    if (gethostname(hostInfo, hostInfoSize-1) == 0) {
+      hostInfo[hostInfoSize-1] = '\0'; //To prevent overflow
```
Comment (gilbertlee-amd) [inline src/init.cc]: likely cleaner to just initialize both hostInfo and libPathInfo to {} on line 701

File: src/init.cc [hunk 7]
```diff
@@ -683,17 +689,44 @@ static ncclResult_t devCommSetup(ncclComm_t comm) {
...
+    } else {
+      // Returns Unknown in hostInfo if function call unsuccessful
+      strncpy(hostInfo, "Unknown", hostInfoSize);
+    }
+    
+    // Retrieve librccl path
+    Dl_info pathInfo;
+    if (dladdr((void*)ncclCommInitRank, &pathInfo)) {
+      strncpy(libPathInfo, pathInfo.dli_fname, libPathInfoSize-1);
+      libPathInfo[libPathInfoSize-1] = '\0'; //To prevent overflow
```
Comment (gilbertlee-amd) [inline src/init.cc]: See comment above

File: src/init.cc [hunk 8]
```diff
@@ -683,17 +689,40 @@ static ncclResult_t devCommSetup(ncclComm_t comm) {
...
   static int shown = 0;
   if (shown == 0 && ncclDebugLevel >= NCCL_LOG_VERSION) {
-    printf("%s %s\n", VERSION_STRING, rcclGitHash);
+    char hostInfo[HOST_NAME_MAX] = {}, libPathInfo[2048] = {};
+    size_t hostInfoSize = sizeof(hostInfo), libPathInfoSize = sizeof(libPathInfo);
+    
+    // Retrieve Hostname info
+    if (gethostname(hostInfo, hostInfoSize-1) != 0) {
+      // Returns Unknown in hostInfo if function call unsuccessful
+      strncpy(hostInfo, "Unknown", hostInfoSize-1);
```
Comment (rahulvaidya20) [inline src/init.cc:707]: @corey-derochie-amd Initialized char arrays to NULL on line 701. So not appending a NULL char at the end for hostInfo and libPathInfo.

**Pull request merged**